### ISSUE TITLE
Relax name character restrictions

### ISSUE THREAD

**Issue by aaronhelton** [RDFLib/rdflib#742]
As of the latest XML 1.0 Recommendation (Fifth Edition, 26 November 2008, referenced in namespace.py), the restrictions on name characters have been significantly relaxed, but RDFLib still enforces the restrictions outlined as of Fourth Edition (16 August 2006). 

In Fourth Edition, this URI is well-formed: http://some.uri/foo/C12345
But this one is not: http://some.uri/foo/12345

These are both valid URIs in Fifth Edition, which a number of free and commercial store providers support or (at least implicitly) require by virtue of implementation.

What are the implications to the project for supporting the loosened restrictions issued in Fifth Edition of the Recommendation, namely allowing start characters to be Unicode category "Nd"? Without that support, RDFLib will be incompatible with the kinds of triple store providers that implement patterns such as [base url]/[project]/[incremental ID] where [incremental ID] is numeric (Unicode category Nd).

Here is a relevant quote from the latest version of the Recommendation:
> This document is a W3C Recommendation. This fifth edition is not a new version of XML. As a convenience to readers, it incorporates the changes dictated by the accumulated errata (available at http://www.w3.org/XML/xml-V10-4e-errata) to the Fourth Edition of XML 1.0, dated 16 August 2006. In particular, erratum [E09] relaxes the restrictions on element and attribute names, thereby providing in XML 1.0 the major end user benefit currently achievable only by using XML 1.1. As a consequence, many possible documents which were not well-formed according to previous editions of this specification are now well-formed, and previously invalid documents using the newly-allowed name characters in, for example, ID attributes, are now valid.

**Comment by joernhees** [comment]
i'm not sure what / where the problem is, maybe you can point us to it?

```python
In [1]: from rdflib import *
RDFLib Version: 5.0.0-dev

In [2]: URIRef('http://some.uri/foo/12345')
Out[2]: rdflib.term.URIRef(u'http://some.uri/foo/12345')

In [3]: URIRef('http://some.uri/foo/12345').n3()
Out[3]: u'<http://some.uri/foo/12345>'

In [4]: Namespace('http://some.uri/foo/')
Out[4]: Namespace(u'http://some.uri/foo/')

In [5]: ns = Namespace('http://some.uri/foo/')

In [6]: ns['12345']
Out[6]: rdflib.term.URIRef(u'http://some.uri/foo/12345')
```

**Comment by gromgull** [comment]
@joernhees it's probably about XML serialisation?

**Comment by aaronhelton** [comment]
It has nothing to do with serialization, at least not directly. namespace.py is implementing the XML specification in the function "is_ncname()", where you're checking to see if the entity name is valid, character by character.

Here's the change I made in my local fork for reference: https://github.com/aaronhelton/rdflib/commit/6aabdf794c037f1ff18f9e2c274617da993fb53f

If "Nd" isn't available as a NAME_START_CATEGORY, then creating a Resource that starts with a numeric identifier fails. This restriction was deprecated in the 5th edition of the W3C Recommendation, and it turns out I need to be able to start my entity names with a number. Out of the box, this fails in RDFLib.

Please let me know if you need additional clarification. I've made the change on my copy, and it works for the use cases I have, but I don't know what the implications are library-wide.

**Comment by joernhees** [comment]
is that problem surfacing anywhere?

or in other words: could you provide a test that fails at the moment and doesn't after your fix?

**Comment by aaronhelton** [comment]
The function that fails here is split_uri from namsepace.py

`>>> from rdflib import *
INFO:rdflib:RDFLib Version: 4.2.2
>>> from rdflib.namespace import split_uri
>>> ns = Namespace('http://some.uri/foo/')
>>> t = ns['1234']
>>> split_uri(t)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ec2-user/projects/venv_rdflib_test/local/lib/python3.4/site-packages/rdflib/namespace.py", line 502, in split_uri
    raise Exception("Can't split '%s'" % uri)
Exception: Can't split 'http://some.uri/foo/1234'
>>> t
rdflib.term.URIRef('http://some.uri/foo/1234')`

The function is necessary for the Graph to run the qname function, which ends up calling split_uri via the compute_qname function:
`>>> from rdflib.namespace import NamespaceManager
>>> g = Graph()
>>> nsm = NamespaceManager(g)
>>> nsm.qname(t)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ec2-user/projects/venv_rdflib_test/local/lib/python3.4/site-packages/rdflib/namespace.py", line 294, in qname
    prefix, namespace, name = self.compute_qname(uri)
  File "/home/ec2-user/projects/venv_rdflib_test/local/lib/python3.4/site-packages/rdflib/namespace.py", line 330, in compute_qname
    namespace, name = split_uri(uri)
  File "/home/ec2-user/projects/venv_rdflib_test/local/lib/python3.4/site-packages/rdflib/namespace.py", line 502, in split_uri
    raise Exception("Can't split '%s'" % uri)
Exception: Can't split 'http://some.uri/foo/1234'
`

This works fine after I make the change linked upthread:
`>>> from rdflib import *
INFO:rdflib:RDFLib Version: 4.2.2
>>> from rdflib.namespace import split_uri, NamespaceManager
>>> ns = Namespace('http://some.uri/foo/')
>>> t = ns['1234']
>>> g = Graph()
>>> nsm = NamespaceManager(g)
>>> nsm.qname(t)
'ns1:1234'
`

Hope that helps.

**Comment by delocalizer** [comment]
This issue is resolved since release of 5.0.0 with the addition of `Nd` category to `rdflib.namespace.SPLIT_START_CATEGORIES`
The example given above that failed in 4.2.2 works now.
